### ISSUE TITLE
 Questions about performance of focal loss and label smoothing

### ISSUE THREAD

**Issue by kimkihoon0515** [ultralytics/yolov5#8469]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I tried two methods to improve performance. 
However, the performance was the best when applied as a baseline. 
Can you roughly know for what reason?

**Comment by glenn-jocher** [comment]
@kimkihoon0515 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by kimkihoon0515** [comment]
@glenn-jocher thx for the brief response. Can u tell me why u think those methods didn't work well on your training environments?

**Comment by buxihuo** [comment]
> @glenn-jocher thx for the brief response. Can u tell me why u think those methods didn't work well on your training environments?

@kimkihoon0515 Hello, I have some ideas about these two technologies. I wrote the answer with translation software. Focal loss can alleviate the imbalance of foreground background proportion in single-stage object detection, while the classification loss of yolov5 is only calculated on foreground objects, and the object loss is not seriously unbalanced. Moreover, focal has some parameters, so it is not as good as BCE loss function (baseline). While the label smoothing technology improves the robustness, it will cause the performance growth to slow down in the early stage of training, and will lose in class information. It also needs to adjust the degree of smoothing. Different values need to be adjusted for different data sets to obtain better results, and the growth effect is not great.

**Comment by ckyrkou** [comment]
> alleviate the imbalance of foreground background proportion in single-stage object detection, while the classification loss of yolov5 is only calculated on foreground objects, and the object loss is not seriously unbalance

I am curious why you say that the objectness loss in not imbalanced. A detection layer can have size of 26x26 and only a few objects say 10, so it could be in the range of 10:670. Since the majority of these locations are non-objects then the output will tend to be zero, hence no object are detected.

**Comment by glenn-jocher** [comment]
@ckyrkou Good point! The output distribution could indeed become imbalanced, especially for the small objects you mentioned. Focal loss is designed to address this issue by down-weighting the easy examples. However, in YOLOv5, the default settings and anchor box system tend to balance the object/background ratio, often at the pixel level. This means that in practice, the objectness loss in YOLOv5 is often balanced internally. While focal loss could still be beneficial in certain scenarios, the default balance commonly minimizes the need for it. Thanks for bringing up the discussion! Let me know if there's anything else you'd like to explore.

**Comment by ckyrkou** [comment]
Thanks for the reply so it turns out that the issues I spent two years on trying to identify between my efforts to reproduce yolo with custom code and the yolov5 repo was due to the facts that:
1) The default display threshold for bounding boxes is 0.25 not the one used during training
2) The map evaluation uses a confidence threshold of 0.001

So the code I was writing was correct but I could never match the same results because the evaluation protocol was different and it didn't even cross my mind that these parameters would be hard coded in such a way.
For the map I have neve seen this threshold used before. 

Anyway it was a fun experience and chance to dig deep into the yolov5 codebase and appreciate how well written it is :)

**Comment by glenn-jocher** [comment]
@ckyrkou that's a huge discovery! It's fascinating how such subtle differences can have significant impacts on results. 🕵️‍♂️ Thanks for sharing your insights, and we're glad to hear that your deep dive into the YOLOv5 codebase was a rewarding experience. 🚀 Should you have any further questions or insights to share, feel free to reach out. Keep up the great work!

**Comment by ckyrkou** [comment]
Well not so huge discovery I would say. But having a reference to the correct way to compute mAP evaluation would be nice. For instance how is this done in the COCO evaluator?

**Comment by glenn-jocher** [comment]
@ckyrkou the COCO evaluator computes mAP by considering all bounding boxes with a confidence score above a certain threshold (default 0.001). The evaluation is performed by comparing the predicted bounding boxes with the ground truth boxes using different Intersection over Union (IoU) thresholds, typically ranging from 0.5 to 0.95.

You can find more details on how the COCO evaluator computes mAP and handles confidence thresholds in the official COCO repository and the associated documentation. This could serve as a valuable reference for understanding the evaluation process. If you have any further questions or need additional insights, feel free to ask.

**Comment by ckyrkou** [comment]
Great to know. Thanks for the prompt response.

**Comment by ckyrkou** [comment]
I would appreciate if you can pinpoint me to the exact place in the COCO repository and the associated documentation where these are mentioned.

**Comment by glenn-jocher** [comment]
@ckyrkou apologies for any confusion, but there seems to have been a misunderstanding. The details I referred to aren't directly documented in the COCO repository, as the specific implementation details can vary across different frameworks and repositories. For insights on the COCO evaluation specifics, I recommend referring to the official COCO dataset documentation and exploring related research papers for a deeper understanding.

If you have any more questions or need further assistance, feel free to ask!
